File: src/conversation_history.py
"""
Conversation History Module
Stores and retrieves conversation history for context-aware AI responses
"""
import json
from datetime import datetime
from typing import List, Dict, Optional


class ConversationHistory:
    """Manages conversation history with timestamps and metadata"""
    
    def __init__(self, max_history: int = 50):
        """
        Initialize conversation history
        
        Args:
            max_history: Maximum number of messages to keep in memory
        """
        self.messages: List[Dict] = []
        self.max_history = max_history
        self.session_id = self._generate_session_id()
        
    def _generate_session_id(self) -> str:
        """Generate unique session ID"""
        return datetime.now().strftime("%Y%m%d_%H%M%S")
    
    def add_message(self, role: str, content: str, safety_flag: Optional[str] = None) -> None:
        """
        Add a message to history
        
        Args:
            role: 'user' or 'assistant'
            content: Message content
            safety_flag: Optional flag if message was flagged for safety concerns
        """
        message = {
            "timestamp": datetime.now().isoformat(),
            "role": role,
            "content": content,
            "safety_flag": safety_flag
        }
        self.messages.append(message)
        
        # Keep only recent messages if exceeding max_history
        if len(self.messages) > self.max_history:
            self.messages.pop(0)
    
    def get_recent_context(self, num_messages: int = 10) -> List[Dict]:
        """
        Get recent messages for context
        
        Args:
            num_messages: Number of recent messages to retrieve
            
        Returns:
            List of recent messages
        """
        return self.messages[-num_messages:] if self.messages else []
    
    def get_user_messages(self) -> List[str]:
        """Get all user messages from history"""
        return [msg["content"] for msg in self.messages if msg["role"] == "user"]
    
    def get_flagged_messages(self) -> List[Dict]:
        """Get all messages that were flagged for safety concerns"""
        return [msg for msg in self.messages if msg["safety_flag"] is not None]
    
    def get_session_summary(self) -> Dict:
        """Get summary of current session"""
        return {
            "session_id": self.session_id,
            "total_messages": len(self.messages),
            "user_messages": len([m for m in self.messages if m["role"] == "user"]),
            "assistant_messages": len([m for m in self.messages if m["role"] == "assistant"]),
            "flagged_count": len(self.get_flagged_messages()),
            "start_time": self.messages[0]["timestamp"] if self.messages else None,
            "end_time": self.messages[-1]["timestamp"] if self.messages else None
        }
    
    def save_to_file(self, filepath: str) -> None:
        """Save conversation history to JSON file"""
        with open(filepath, 'w') as f:
            json.dump({
                "session_id": self.session_id,
                "messages": self.messages,
                "summary": self.get_session_summary()
            }, f, indent=2)
    
    def load_from_file(self, filepath: str) -> None:
        """Load conversation history from JSON file"""
        try:
            with open(filepath, 'r') as f:
                data = json.load(f)
                self.messages = data.get("messages", [])
                self.session_id = data.get("session_id", self._generate_session_id())
        except FileNotFoundError:
            print(f"History file not found: {filepath}")
    
    def clear_history(self) -> None:
        """Clear all messages from history"""
        self.messages = []
    
    def get_all_messages(self) -> List[Dict]:
        """Get all messages from history"""
        return self.messages.copy()

File: javascript/src/ConversationHistory.js
/**
 * Conversation History Module
 * Stores and retrieves conversation history for context-aware AI responses
 */

class ConversationHistory {
  /**
   * Initialize conversation history
   * @param {number} maxHistory - Maximum number of messages to keep in memory
   */
  constructor(maxHistory = 50) {
    this.messages = [];
    this.maxHistory = maxHistory;
    this.sessionId = this._generateSessionId();
  }

  /**
   * Generate unique session ID
   * @returns {string}
   */
  _generateSessionId() {
    const now = new Date();
    const year = now.getFullYear();
    const month = String(now.getMonth() + 1).padStart(2, '0');
    const day = String(now.getDate()).padStart(2, '0');
    const hours = String(now.getHours()).padStart(2, '0');
    const minutes = String(now.getMinutes()).padStart(2, '0');
    const seconds = String(now.getSeconds()).padStart(2, '0');
    return `${year}${month}${day}_${hours}${minutes}${seconds}`;
  }

  /**
   * Add a message to history
   * @param {string} role - 'user' or 'assistant'
   * @param {string} content - Message content
   * @param {string|null} safetyFlag - Optional flag if message was flagged
   */
  addMessage(role, content, safetyFlag = null) {
    const message = {
      timestamp: new Date().toISOString(),
      role,
      content,
      safetyFlag
    };
    this.messages.push(message);

    // Keep only recent messages if exceeding max_history
    if (this.messages.length > this.maxHistory) {
      this.messages.shift();
    }
  }

  /**
   * Get recent messages for context
   * @param {number} numMessages - Number of recent messages to retrieve
   * @returns {Array}
   */
  getRecentContext(numMessages = 10) {
    return this.messages.length > 0
      ? this.messages.slice(-numMessages)
      : [];
  }

  /**
   * Get all user messages from history
   * @returns {Array}
   */
  getUserMessages() {
    return this.messages
      .filter(msg => msg.role === 'user')
      .map(msg => msg.content);
  }

  /**
   * Get all messages that were flagged for safety concerns
   * @returns {Array}
   */
  getFlaggedMessages() {
    return this.messages.filter(msg => msg.safetyFlag !== null);
  }

  /**
   * Get summary of current session
   * @returns {object}
   */
  getSessionSummary() {
    const userMessages = this.messages.filter(m => m.role === 'user').length;
    const assistantMessages = this.messages.filter(m => m.role === 'assistant').length;
    const flaggedCount = this.getFlaggedMessages().length;

    return {
      sessionId: this.sessionId,
      totalMessages: this.messages.length,
      userMessages,
      assistantMessages,
      flaggedCount,
      startTime: this.messages.length > 0 ? this.messages[0].timestamp : null,
      endTime: this.messages.length > 0 ? this.messages[this.messages.length - 1].timestamp : null
    };
  }

  /**
   * Save conversation history to JSON
   * @param {string} filepath - File path to save to
   */
  saveToFile(filepath) {
    const fs = require('fs');
    const data = {
      sessionId: this.sessionId,
      messages: this.messages,
      summary: this.getSessionSummary()
    };
    fs.writeFileSync(filepath, JSON.stringify(data, null, 2));
  }

  /**
   * Load conversation history from JSON file
   * @param {string} filepath - File path to load from
   */
  loadFromFile(filepath) {
    try {
      const fs = require('fs');
      const data = JSON.parse(fs.readFileSync(filepath, 'utf8'));
      this.messages = data.messages || [];
      this.sessionId = data.sessionId || this._generateSessionId();
    } catch (error) {
      console.log(`History file not found: ${filepath}`);
    }
  }

  /**
   * Clear all messages from history
   */
  clearHistory() {
    this.messages = [];
  }

  /**
   * Get all messages from history
   * @returns {Array}
   */
  getAllMessages() {
    return [...this.messages];
  }
}

module.exports = ConversationHistory;
